src/ipred.rs

```diff
@@ -1198,87 +1198,82 @@ unsafe fn filter_fn(
     p6: c_int,
 ) -> c_int {
     if cfg!(any(target_arch = "x86", target_arch = "x86_64")) {
-        *flt_ptr.offset(0) as c_int * p0
-            + *flt_ptr.offset(1) as c_int * p1
-            + *flt_ptr.offset(16) as c_int * p2
-            + *flt_ptr.offset(17) as c_int * p3
-            + *flt_ptr.offset(32) as c_int * p4
-            + *flt_ptr.offset(33) as c_int * p5
-            + *flt_ptr.offset(48) as c_int * p6
+        flt_ptr[0] as c_int * p0
+            + flt_ptr[1] as c_int * p1
+            + flt_ptr[16] as c_int * p2
+            + flt_ptr[17] as c_int * p3
+            + flt_ptr[32] as c_int * p4
+            + flt_ptr[33] as c_int * p5
+            + flt_ptr[48] as c_int * p6
     } else {
-        *flt_ptr.offset(0) as c_int * p0
-            + *flt_ptr.offset(8) as c_int * p1
-            + *flt_ptr.offset(16) as c_int * p2
-            + *flt_ptr.offset(24) as c_int * p3
-            + *flt_ptr.offset(32) as c_int * p4
-            + *flt_ptr.offset(40) as c_int * p5
-            + *flt_ptr.offset(48) as c_int * p6
+        flt_ptr[0] as c_int * p0
+            + flt_ptr[8] as c_int * p1
+            + flt_ptr[16] as c_int * p2
+            + flt_ptr[24] as c_int * p3
+            + flt_ptr[32] as c_int * p4
+            + flt_ptr[40] as c_int * p5
+            + flt_ptr[48] as c_int * p6
     }
 }
 
 cfg_if! {
     if #[cfg(any(target_arch = "x86", target_arch = "x86_64"))] {
-        const FLT_INCR: isize = 2;
+        const FLT_INCR: usize = 2;
     } else {
-        const FLT_INCR: isize = 1;
+        const FLT_INCR: usize = 1;
     }
 }
 
 unsafe fn ipred_filter_rust<BD: BitDepth>(
     mut dst: *mut BD::Pixel,
     stride: ptrdiff_t,
-    topleft_in: *const BD::Pixel,
+    topleft_in: &[BD::Pixel; SCRATCH_EDGE_LEN],
+    topleft_off: usize,
     width: c_int,
     height: c_int,
-    mut filt_idx: c_int,
+    filt_idx: c_int,
     _max_width: c_int,
     _max_height: c_int,
     bd: BD,
 ) {
-    filt_idx &= 511 as c_int;
-    if !(filt_idx < 5) {
-        unreachable!();
-    }
-    let filter: *const i8 = (dav1d_filter_intra_taps[filt_idx as usize]).as_ptr();
-    let mut top: *const BD::Pixel = &*topleft_in.offset(1) as *const BD::Pixel;
-    let mut y = 0;
-    while y < height {
-        let mut topleft: *const BD::Pixel = &*topleft_in.offset(-y as isize) as *const BD::Pixel;
-        let mut left: *const BD::Pixel =
-            &*topleft.offset(-(1 as c_int) as isize) as *const BD::Pixel;
-        let mut left_stride: ptrdiff_t = -(1 as c_int) as ptrdiff_t;
-        let mut x = 0;
-        while x < width {
+    let width = width as usize;
+    let height = height as usize;
+    let stride = BD::pxstride(stride);
+    let filt_idx = filt_idx & 511;
+    assert!(filt_idx < 5);
+
+    let filter = &dav1d_filter_intra_taps[filt_idx as usize];
+    let mut top = topleft_in.as_ptr().add(topleft_off + 1);
+    for y in (0..height).step_by(2) {
+        let mut topleft = topleft_in.as_ptr().add(topleft_off - y);
```

Comment:
```suggestion
        let top = slice::from_raw_parts_mut(top, 4);
        let mut topleft = topleft_in.as_ptr().add(topleft_off - y);
```

Reply:
I had to create the slice from within the `for x` loop because `top` is updated at the end of the loop (not just the `for y` loop), but otherwise this works 👍